input/listen.go

```diff
@@ -8,126 +8,161 @@ import (
+	udpConn *net.UDPConn
+	handler Handler
+}
+
+func NewListener(handler Handler) *Listener {
+	return &Listener{
+		handler: handler,
+	}
+}
+
+func (l *Listener) listen(addr string) error {
+	l.startStoppable()
+
 	// listeners are set up outside of accept* here so they can interrupt startup
-	listener, err := listenTcp(addr)
+	err := l.listenTcp(addr)
 	if err != nil {
 		return err
 	}
 
-	udp_conn, err := listenUdp(addr)
+	err = l.listenUdp(addr)
 	if err != nil {
 		return err
 	}
 
-	go acceptTcp(addr, listener, handler)
-	go acceptUdp(addr, udp_conn, handler)
+	l.wg.Add(2)
+	go l.accept(addr, "tcp", l.acceptTcp, l.listenTcp, l.tcpList)
+	go l.accept(addr, "udp", l.consumeUdp, l.listenUdp, l.udpConn)
 
 	return nil
 }
 
-func listenTcp(addr string) (*net.TCPListener, error) {
-	laddr, err := net.ResolveTCPAddr("tcp", addr)
-	if err != nil {
-		return nil, err
-	}
-	return net.ListenTCP("tcp", laddr)
-}
+func (l *Listener) accept(addr, proto string, consume func(string), reconnect func(string) error, listener Closable) {
+	defer l.wg.Done()
 
-func acceptTcp(addr string, listener *net.TCPListener, handler Handler) {
-	var err error
-	l := listener
 	backoffCounter := &backoff.Backoff{
 		Min: 500 * time.Millisecond,
 		Max: time.Minute,
 	}
 
-	for {
-		log.Notice("listening on %v/tcp", addr)
+	go func() {
+		<-l.shutdown
+		log.Info("shutting down %v/%s, closing socket", addr, proto)
+		listener.Close()
+	}()
 
-		for {
-			// wait for a tcp connection
-			c, err := l.AcceptTCP()
-			if err != nil {
-				log.Error("error accepting on %v/tcp, closing connection: %s", addr, err)
-				l.Close()
-				break
+	for {
+		log.Notice("listening on %v/%s", addr, proto)
+
+		consume(addr)
+
+		select {
+		case <-l.shutdown:
+			return
+		default:
+			for {
+				log.Notice("reopening %v/%s", addr, proto)
+				err := reconnect(addr)
+				if err == nil {
+					backoffCounter.Reset()
+					break
+				}
+
+				backoffDuration := backoffCounter.Duration()
+				log.Error("error listening on %v/%s, retrying after %v: %s", addr, proto, backoffDuration, err)
+				select {
+				case <-l.shutdown:
+					log.Info("shutting down %v/%s, closing socket", addr, proto)
+					return
+				case <-time.After(backoffDuration):
+					// retry to connect
+				}
 			}
-
-			// handle the connection
-			log.Debug("listen.go: tcp connection from %v", c.RemoteAddr())
-			go acceptTcpConn(c, handler)
 		}
+	}
+}
 
-		for {
-			log.Notice("reopening %v/tcp", addr)
+func (l *Listener) listenTcp(addr string) error {
+	laddr, err := net.ResolveTCPAddr("tcp", addr)
+	if err != nil {
+		return err
+	}
+	l.tcpList, err = net.ListenTCP("tcp", laddr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-			l, err = listenTcp(addr)
-			if err == nil {
-				backoffCounter.Reset()
-				break
+func (l *Listener) acceptTcp(addr string) {
+	for {
+		c, err := l.tcpList.AcceptTCP()
+		if err != nil {
+			select {
+			// socket has been closed due to shut down, log and return
```

Comment:
this comment is incorrect. and if it were correct it would be redundent, so just remove it.